File: java/src/main/java/sk/epostak/sdk/DuplicateInvoiceNumberException.java
package sk.epostak.sdk;

import java.util.List;

/**
 * Thrown when {@code POST /api/v1/documents/send} (or the dashboard
 * create endpoint) rejects an outbound invoice whose {@code invoice_number}
 * already exists for the firm.
 * <p>
 * The conflict key is {@code (firmId, invoiceNumber)} — the recipient is
 * intentionally NOT part of it; outbound numbering belongs to the sender.
 *
 * <pre>{@code
 * try {
 *     client.documents().send(req);
 * } catch (DuplicateInvoiceNumberException e) {
 *     ExistingDocument existing = e.getExistingDocument();
 *     if (existing != null) {
 *         System.err.println("Already sent at " + existing.sentAt() +
 *                            ", id=" + existing.id());
 *     }
 * }
 * }</pre>
 */
public class DuplicateInvoiceNumberException extends EPostakException {

    /**
     * Identification of the recipient on the existing duplicate invoice.
     * Each field can be {@code null} if the original invoice did not have
     * it stored — in particular {@code peppolId} is always present when
     * the parent {@code recipient} object is non-null, but {@code ico} and
     * {@code name} are optional on the API side.
     */
    public record Recipient(String peppolId, String ico, String name) {}

    /**
     * The pre-existing outbound invoice that triggered the conflict.
     * {@code sentAt} is an ISO 8601 string — {@code peppolSentAt} if the
     * original was already delivered, otherwise {@code createdAt}.
     * {@code recipient} is {@code null} when the original invoice had no
     * recipient Peppol ID stored.
     */
    public record ExistingDocument(
        String id,
        String invoiceNumber,
        String status,
        String sentAt,
        Recipient recipient
    ) {}

    private final List<String> conflictKey;
    private final ExistingDocument existingDocument;

    public DuplicateInvoiceNumberException(
            int status,
            String message,
            String code,
            Object details,
            String type,
            String title,
            String detail,
            String instance,
            String requestId,
            String requiredScope,
            List<String> conflictKey,
            ExistingDocument existingDocument
    ) {
        super(status, message, code, details, type, title, detail, instance, requestId, requiredScope);
        this.conflictKey = conflictKey == null
            ? List.of("firmId", "invoiceNumber")
            : List.copyOf(conflictKey);
        this.existingDocument = existingDocument;
    }

    /** Always {@code ["firmId", "invoiceNumber"]}. */
    public List<String> getConflictKey() {
        return conflictKey;
    }

    /**
     * The pre-existing outbound invoice that caused the conflict, or
     * {@code null} if it was deleted between the constraint hit and the
     * server-side lookup.
     */
    public ExistingDocument getExistingDocument() {
        return existingDocument;
    }
}

File: dotnet/src/EPostak/DuplicateInvoiceNumberException.cs
namespace EPostak;

/// <summary>
/// Recipient identification on the existing duplicate invoice. Any field
/// can be <c>null</c> if the original invoice did not have it stored.
/// </summary>
public sealed record DuplicateInvoiceRecipient(string? PeppolId, string? Ico, string? Name);

/// <summary>
/// The pre-existing outbound invoice that triggered the conflict.
/// <see cref="SentAt"/> is an ISO 8601 string — <c>peppolSentAt</c> if the
/// original was already delivered, otherwise <c>createdAt</c>.
/// </summary>
public sealed record DuplicateInvoiceExistingDocument(
    string Id,
    string InvoiceNumber,
    string Status,
    string SentAt,
    DuplicateInvoiceRecipient? Recipient
);

/// <summary>
/// Thrown when <c>POST /api/v1/documents/send</c> (or the dashboard create
/// endpoint) rejects an outbound invoice whose <c>invoice_number</c> is
/// already in use for the firm.
/// </summary>
/// <remarks>
/// The conflict key is <c>(firmId, invoiceNumber)</c> — recipient is NOT
/// part of it; outbound numbering belongs to the sender.
/// </remarks>
/// <example>
/// <code>
/// try
/// {
///     await client.Documents.SendAsync(req, ct);
/// }
/// catch (DuplicateInvoiceNumberException ex)
/// {
///     var existing = ex.ExistingDocument;
///     if (existing != null)
///         Console.WriteLine($"Already sent on {existing.SentAt}, id={existing.Id}");
/// }
/// </code>
/// </example>
public class DuplicateInvoiceNumberException : EPostakException
{
    /// <summary>Always <c>["firmId", "invoiceNumber"]</c>.</summary>
    public IReadOnlyList<string> ConflictKey { get; }

    /// <summary>
    /// The pre-existing outbound invoice that caused the conflict, or
    /// <c>null</c> if it was deleted between the constraint hit and the
    /// server-side lookup.
    /// </summary>
    public DuplicateInvoiceExistingDocument? ExistingDocument { get; }

    public DuplicateInvoiceNumberException(
        int status,
        string message,
        string? code,
        object? details,
        string? type,
        string? title,
        string? detail,
        string? instance,
        string? requestId,
        string? requiredScope,
        IReadOnlyList<string>? conflictKey,
        DuplicateInvoiceExistingDocument? existingDocument
    ) : base(status, message, code, details, type, title, detail, instance, requestId, requiredScope)
    {
        ConflictKey = conflictKey ?? Array.AsReadOnly(new[] { "firmId", "invoiceNumber" });
        ExistingDocument = existingDocument;
    }
}
